futures-util/src/stream/split.rs

```diff
@@ -82,27 +78,31 @@ impl<S: Sink> Sink for SplitSink<S> {
         Ok(())
     }
 
-    fn start_close(&mut self) -> Result<(), S::SinkError> {
-        self.do_close = true;
-        Ok(())
-    }
-
     fn poll_flush(&mut self, cx: &mut task::Context) -> Poll<(), S::SinkError> {
         match self.lock.poll_lock(cx) {
             Async::Ready(mut inner) => {
                 if self.slot.is_some() {
                     try_ready!(inner.poll_ready(cx));
                     inner.start_send(self.slot.take().unwrap())?;
                 }
-                if self.do_close {
-                    inner.start_close()?;
-                    self.do_close = false;
-                }
                 inner.poll_flush(cx)
             }
             Async::Pending => Ok(Async::Pending),
         }
     }
+
+    fn poll_close(&mut self, cx: &mut task::Context) -> Poll<(), S::SinkError> {
+        match self.lock.poll_lock(cx) {
+            Async::Ready(mut inner) => {
+                if self.slot.is_some() {
```

Comment:
tiny thing, but this can be rewritten as `if let Some(val) = self.slot.take()`

Reply:
Unfortunately not-- the try_ready! may cause an early return, in which case we don't want to remove the item from the slot.